src/overlay/OverlayManagerImpl.cpp

```diff
@@ -340,10 +387,30 @@ OverlayManagerImpl::updateSizeCounters()
 }
 
 void
-OverlayManagerImpl::addPendingPeer(Peer::pointer peer)
+OverlayManagerImpl::addInboundConnection(Peer::pointer peer)
 {
+    assert(peer->getRole() == Peer::REMOTE_CALLED_US);
+
     if (mShuttingDown ||
```

Comment:
I wonder if we should always allow preferred peers to be added (even if it means going a little bit over `MAX_PENDING_CONNECTIONS`) ?

Reply:
I'm not sure if we know at this point if peer is preferred.
At this point no messages between peers were exchanged. So we do not know the KEY and cannot check by PREFERRED_PEER_KEYS.

But we also haven't received listeningPort, so we cannot really check if this peer is on our PREFERRED_PEERS list. Or we can just search by IP here...

Reply:
yeah I am thinking preferred by IP

Reply:
So we should ignore port? For example if we prefer address "127.0.0.1:1000" and peer connects from "127.0.01", but we don't know its listening port, should we accept it anyway?

Reply:
yes at this stage it's safe to ignore the port when accepting connections as preferred peers are really trusted by ip (we trust the port number supplied). The later check done in `acceptAuthenticatedPeer` will ensure that there is a single connection to that ip.